// RUN: %clang_cc1 -triple=x86_64-unknown-unknown -frecovery-ast -frecovery-ast-type -o - %s -fsyntax-only -verify

namespace NoCrash{
struct Indestructible {
  // Indestructible();
  ~Indestructible() = delete; // expected-note {{deleted}}
};
Indestructible make_indestructible();

// no crash on HasSideEffect.
void test() {
  int s = sizeof(make_indestructible()); // expected-error {{deleted}}
}
}

namespace {
void foo(); // expected-note {{requires 0 arguments}}
class Y {
  // no "field has incomplete type" diagnostic.
  typeof(foo(42)) invalid; // expected-error {{no matching function}}
};
}
